### PR TITLE
normaliser: simplified events => compiled event

### PR DESCRIPTION
**Pull request by Gorash**

### CODE REVIEW

File: src/core/utils/EventNormalizer.ts [hunk 1]
```diff
@@ -152,24 +155,16 @@ export class EventNormalizer {
      * @private
      * @returns {CompiledEvent}
      */
-    _compileEvents(): CompiledEvent {
-        if (this._compiledEvent) {
-            return this._compiledEvent;
+    _addSimplifiedEvent(ev: CompiledEvent): void {
```
Comment (dmo-odoo) [inline src/core/utils/EventNormalizer.ts]: At this point, I wonder what is the point of CompiledEvent. Couldn't we just use Event here ?

File: src/core/utils/EventNormalizer.ts [hunk 2]
```diff
@@ -152,24 +155,16 @@ export class EventNormalizer {
...
+        if (!this._simplifiedEvents) {
+            this._simplifiedEvents = {
+                events: [],
+                clone: undefined,
```
Comment (dmo-odoo) [inline src/core/utils/EventNormalizer.ts]: `false`

File: src/core/utils/EventNormalizer.ts [hunk 3]
```diff
@@ -58,7 +58,7 @@ interface EventListenerDeclaration {
 export class EventNormalizer {
     editable: HTMLElement;
     _eventListeners: EventListenerDeclaration[] = [];
-    _compiledEvent: CompiledEvent;
+    _simplifiedEvents: { events: CompiledEvent[]; clone: ClonedNode };
```
Comment (dmo-odoo) [inline src/core/utils/EventNormalizer.ts]: What is "simplified" here ? Why not `eventsQueue` or something like that ?

File: src/core/utils/EventNormalizer.ts [hunk 4]
```diff
@@ -207,34 +202,91 @@ export class EventNormalizer {
...
+     * Call _processCompiledEvents who process this compiled event.
      *
      * @private
      */
-    _processCompiledEvent(): void {
+    _processSimplifiedEvents(): void {
         // Store and reset compiled event for the next stack.
-        const compiledEvent = this._compiledEvent;
-        this._compiledEvent = null;
+        const simplifiedEvents = this._simplifiedEvents;
```
Comment (dmo-odoo) [inline src/core/utils/EventNormalizer.ts]: In ES6, wouldn't it be possible to do `const {events, clone} = this._simplifiedEvents` ?

File: src/core/utils/EventNormalizer.ts [hunk 5]
```diff
@@ -207,34 +202,91 @@ export class EventNormalizer {
...
+        this._simplifiedEvents = null;
+        const events = simplifiedEvents.events;
+        // const elements = this._processMutationEvent(events);
```
Comment (dmo-odoo) [inline src/core/utils/EventNormalizer.ts]: 👀

File: src/core/utils/EventNormalizer.ts [hunk 6]
```diff
@@ -207,34 +202,91 @@ export class EventNormalizer {
...
+                    {
+                        type: 'composition',
+                        clone: simplifiedEvents.clone,
+                        data: event.data,
+                        defaultPrevented: event.defaultPrevented,
+                        elements: this._processMutationEvent(
+                            events.find(event => event.type === 'mutation'),
+                        ),
+                    },
+                ];
```
Comment (dmo-odoo) [inline src/core/utils/EventNormalizer.ts]: For me, this is a `CompiledEvent`, but the previous one were not.

File: src/core/utils/EventNormalizer.ts [hunk 7]
```diff
@@ -207,34 +202,91 @@ export class EventNormalizer {
...
+                    compiledEvent.defaultPrevented || event.defaultPrevented;
+            }
+            if (event.type === 'mutation') {
+                compiledEvent.elements = this._processMutationEvent(event);
+            }
         }
 
-        // Gather all modified nodes to notify the listener.
+        console.log(simplifiedEvents);
+        console.log(compiledEvents);
```
Comment (dmo-odoo) [inline src/core/utils/EventNormalizer.ts]: 👀

File: src/core/utils/EventNormalizer.ts [hunk 8]
```diff
@@ -207,34 +202,91 @@ export class EventNormalizer {
...
+                    compiledEvent.defaultPrevented || event.defaultPrevented;
+            }
+            if (event.type === 'input') {
+                compiledEvent.data = event.data;
+                compiledEvent.defaultPrevented =
+                    compiledEvent.defaultPrevented || event.defaultPrevented;
+            }
+            if (event.type === 'mutation') {
+                compiledEvent.elements = this._processMutationEvent(event);
+            }
```
Comment (dmo-odoo) [inline src/core/utils/EventNormalizer.ts]: This event type is new, isn't it ?

File: src/core/utils/EventNormalizer.ts [hunk 9]
```diff
@@ -243,7 +295,23 @@ export class EventNormalizer {
...
+        return elements;
+    }
+    /**
+     * Process the compiled event by triggering the corresponding events on the
+     * listener and reset the compiling process for the next stack.
+     *
+     * @private
+     */
+    _processCompiledEvents(compiledEvent: CompiledEvent): void {
+        // TODO: transfer defaultPrevented information
```
Comment (dmo-odoo) [inline src/core/utils/EventNormalizer.ts]: Hasn't this been done above ?

File: src/core/utils/EventNormalizer.ts [hunk 10]
```diff
@@ -777,66 +842,47 @@ export class EventNormalizer {
         // See comment on the same line in _onCompositionStart handler.
         if (this.editable.style.display === 'none') return;
 
-        const compiledEvent = this._compileEvents();
+        const compiledEvent = {
+            type: null,
+            key: 'Unidentified',
+            data: '',
+            defaultPrevented: ev.defaultPrevented,
+        };
```
Comment (dmo-odoo) [inline src/core/utils/EventNormalizer.ts]: 👀

File: src/core/utils/EventNormalizer.ts [hunk 11]
```diff
@@ -845,22 +891,21 @@ export class EventNormalizer {
      * @param {KeyboardEvent} ev
      */
     _onKeyDownOrKeyPress(ev: KeyboardEvent): void {
+        console.log(ev.type, ev.key);
```
Comment (dmo-odoo) [inline src/core/utils/EventNormalizer.ts]: 👀

File: src/core/utils/EventNormalizer.ts [hunk 12]
```diff
@@ -927,11 +972,13 @@ export class EventNormalizer {
...
         if (
-            (!this._compiledEvent || this._compiledEvent.key === 'a') &&
+            (!this._simplifiedEvents ||
+                this._simplifiedEvents.events.filter(ev => ev.type === 'keydown' && ev.key === 'a')
+                    .length) &&
             this._isSelectAll(this._getRange())
         ) {
             this._triggerEvent('selectAll', {
-                origin: this._compiledEvent ? 'keypress' : 'pointer',
+                origin: this._simplifiedEvents ? 'keypress' : 'pointer',
```
Comment (dmo-odoo) [inline src/core/utils/EventNormalizer.ts]: Not sure this makes sense anymore

File: src/core/utils/EventNormalizer.ts (hunk 1 above)
Comment (Gorash) [reply src/core/utils/EventNormalizer.ts]: I wanted to see with you, in fact I asked myself the question. Yes, I'm using native event.

**Comment by dmo-odoo** [comment]
- modified keys should not trigger insert if they did not actually insert text (ctrl+a should not insert a)
- shift click from h1 to somewhere in the p triggers setRange with wrong values (click in h1, then hold shift and click in the text)

File: src/core/utils/EventNormalizer.ts [hunk 13]
```diff
@@ -222,79 +265,244 @@ export class EventNormalizer {
...
+        const type = compiledEvent.type;
+        const key = compiledEvent.key;
+        const data = compiledEvent.data;
+        if (type === 'composition') {
+            if (['insertCompositionText', 'insertReplacementText'].includes(ev.inputType)) {
+                // Some spell checking keyboards do not properly fill the textual
+                // data of the composition in the compositionend event. To protect
+                // against this, use the data from the input event instead.
+                compiledEvent.data = ev.data;
+                this._cloneForComposition();
```
Comment (dmo-odoo) [inline src/core/utils/EventNormalizer.ts]: All these `_cloneForComposition`, aren't they too late now that we analyze the events after the tick ? Shouldn't they be in their respective `_on*` event listeners instead ?

Comment (Gorash) [reply src/core/utils/EventNormalizer.ts]: change proposal to fix the corrector on mac. Another solution is to give access to the vDocument in the normalizer

**Comment by Gorash** [comment]
refactoring (with @dmo-odoo and @Goaman ) and add test generator: https://github.com/odoo-dev/jabberwock/pull/97

**Pull request closed without merging**